### PR TITLE
Pack to integrate OpenAI API to Stackstorm

### PR DESCRIPTION
**Pull request by dalarsen**
Pack that adds two actions (chat completion and image generation) through the OpenAI API.  Hopefully I've included most of the necessary stuff for this to be correct but please let me know of any changes that are needed.  Thank you!

### CODE REVIEW

**Comment by armab** [comment]
Great stuff!
Noticed there's a StackStorm logo as `icon.png`. Worth replacing it with the official OpenAI logo.

**Comment by dalarsen** [comment]
> Great stuff! Noticed there's a StackStorm logo as `icon.png`. Worth replacing it with the official OpenAI logo.

Nice catch, thanks; i've replaced the logo.

File: openai/requirements.txt
Comment (armab) [inline openai/requirements.txt:1]: Ideally to pin the version to the current major (interval) in case of the upstream breaking changes

**Review by arm4b — CHANGES_REQUESTED** [review]
Awesome addition to the StackStorm Exchange. Thanks for the contribution!

Looking at other packs, please also include the following to make it complete:
- `CHANGES.md` ([example](https://github.com/StackStorm-Exchange/stackstorm-logicmonitor/blob/main/CHANGES.md))
- `openai.yaml.example` (like [this](https://github.com/StackStorm-Exchange/stackstorm-vault/blob/master/vault.yaml.example)) - just an example of pack config with some dummy data
- `.github/CODEOWNERS` ([example](https://github.com/StackStorm-Exchange/stackstorm-napalm/blob/master/.github/CODEOWNERS)) and include yourself as a maintainer + @StackStorm-Exchange/tsc, - this way on every PR and issue we'll get notifications
- Add yourself as a Maintainer in the README footer ([example format](https://github.com/StackStorm-Exchange/stackstorm-napalm/blob/master/README.md#maintainers))

Some basic [/tests/](https://docs.stackstorm.com/development/pack_testing.html) would be nice if you can add them too, but not required.

I think some ChatOps action aliases would be ideal for such a pack, but we can add them in a later PRs after creating a pack.

**Comment by dalarsen** [comment]
Thanks for the feedback!  I think I've addressed everything, if I've missed anything please let me know.  Thank you!

File: openai/.github/CODEOWNERS
Comment (armab) [inline openai/.github/CODEOWNERS:11]:
```suggestion
* @dalarsen @StackStorm-Exchange/tsc
```

**Comment by armab** [comment]
Thanks!
Will ask if @StackStorm-Exchange/tsc has any feedback, but the PR looks great to me!

I'll need to check why CI is failing on this as well before merging.

File: openai/requirements.txt
Comment (arm4b) [inline openai/requirements.txt]: The CI machinery should be working now and we've reached pack-related CI runs.

It's failing on `requirements.txt` now: https://github.com/StackStorm-Exchange/exchange-incubator/actions/runs/5879130202/job/15942547692?pr=179#step:8:1714

@dalarsen Maybe worth trying the requirements without whitespaces:

```suggestion
openai~=0.27
```
Otherwise, last resort is to go with the old notation:
```
openai>=0.27.0,<1.0.0
```

Comment (dalarsen) [reply openai/requirements.txt]: Weird, it's still failing, but I can't reproduce the error locally at all?

```
(venv) dalarsen@EXPWYFP79391M openai % pip install -r requirements.txt 
Collecting openai<1.0,>=0.27.0
  Using cached openai-0.27.8-py3-none-any.whl (73 kB)
Requirement already satisfied: requests>=2.20 in ./venv/lib/python3.9/site-packages (from openai<1.0,>=0.27.0->-r requirements.txt (line 1)) (2.31.0)
Requirement already satisfied: aiohttp in ./venv/lib/python3.9/site-packages (from openai<1.0,>=0.27.0->-r requirements.txt (line 1)) (3.8.5)
Requirement already satisfied: tqdm in ./venv/lib/python3.9/site-packages (from openai<1.0,>=0.27.0->-r requirements.txt (line 1)) (4.66.1)
Requirement already satisfied: certifi>=2017.4.17 in ./venv/lib/python3.9/site-packages (from requests>=2.20->openai<1.0,>=0.27.0->-r requirements.txt (line 1)) (2023.7.22)
Requirement already satisfied: urllib3<3,>=1.21.1 in ./venv/lib/python3.9/site-packages (from requests>=2.20->openai<1.0,>=0.27.0->-r requirements.txt (line 1)) (2.0.4)
Requirement already satisfied: charset-normalizer<4,>=2 in ./venv/lib/python3.9/site-packages (from requests>=2.20->openai<1.0,>=0.27.0->-r requirements.txt (line 1)) (3.2.0)
Requirement already satisfied: idna<4,>=2.5 in ./venv/lib/python3.9/site-packages (from requests>=2.20->openai<1.0,>=0.27.0->-r requirements.txt (line 1)) (3.4)
Requirement already satisfied: attrs>=17.3.0 in ./venv/lib/python3.9/site-packages (from aiohttp->openai<1.0,>=0.27.0->-r requirements.txt (line 1)) (23.1.0)
Requirement already satisfied: multidict<7.0,>=4.5 in ./venv/lib/python3.9/site-packages (from aiohttp->openai<1.0,>=0.27.0->-r requirements.txt (line 1)) (6.0.4)
Requirement already satisfied: async-timeout<5.0,>=4.0.0a3 in ./venv/lib/python3.9/site-packages (from aiohttp->openai<1.0,>=0.27.0->-r requirements.txt (line 1)) (4.0.3)
Requirement already satisfied: frozenlist>=1.1.1 in ./venv/lib/python3.9/site-packages (from aiohttp->openai<1.0,>=0.27.0->-r requirements.txt (line 1)) (1.4.0)
Requirement already satisfied: aiosignal>=1.1.2 in ./venv/lib/python3.9/site-packages (from aiohttp->openai<1.0,>=0.27.0->-r requirements.txt (line 1)) (1.3.1)
Requirement already satisfied: yarl<2.0,>=1.0 in ./venv/lib/python3.9/site-packages (from aiohttp->openai<1.0,>=0.27.0->-r requirements.txt (line 1)) (1.9.2)
Installing collected packages: openai
Successfully installed openai-0.27.8
WARNING: You are using pip version 21.3.1; however, version 23.2.1 is available.
You should consider upgrading via the '/Users/dalarsen/PycharmProjects/public/exchange-incubator/openai/venv/bin/python -m pip install --upgrade pip' command.
(venv) dalarsen@EXPWYFP79391M openai % cat requirements.txt 
openai>=0.27.0,<1.0
```

I tried it with --no-binary as well to make sure it would still work if there wasn't a locally cached wheel of the thing, and it still worked fine...

File: openai/requirements.txt
Comment (arm4b) [inline openai/requirements.txt:1]: Ah, I see now what's going on.
https://pypi.org/project/openai/ Requires minimum `Python >=3.7.1`, while the tests are ran under the python 3.6

The technical debt of maintaining the Python versions follows us everywhere: https://github.com/StackStorm/community/issues/103

Comment (dalarsen) [reply openai/requirements.txt:1]: Ahh, that explains it!  As a workaround would it be an option for me to just modify the github action that does the tests to use a later Python version (and/or change the image being used to one that's noted to support a later Python version)?  Or is there more to it than that?

Comment (arm4b) [reply openai/requirements.txt:1]: Yeah, I think it worth trying with Python 3.8 👍  as that's the other version supported by StackStorm.

After pack is promoted to Exchange, we use a common GHA for the builds: https://github.com/StackStorm-Exchange/stackstorm-jira/blob/master/.github/workflows/build_and_test.yaml which is ran under the py3.6
But I just discovered some packs (https://github.com/StackStorm-Exchange/stackstorm-vault/blob/master/.github/workflows/build_and_test.yaml) have this workflow overridden, so it should be fine to do that as well in this new pack.

Comment (dalarsen) [reply openai/requirements.txt:1]: I changed the workflow to use 3.8 and it looks like it's passing now, though I'm not sure if I did it in a way that it would be kept if it were to be moved to the Exchange.

**Review by arm4b — APPROVED** [review]
Nice! Thanks @dalarsen!

@cognifloyd I'm going to bootstrap the pack next, following the https://github.com/StackStorm-Exchange/exchange-incubator/pull/172 you've built and see what happens 🤞  Do you foresee any potential issues by switching the CI workflow to py3.8 in this PR?

**Comment by arm4b** [comment]
!bootstrap pack

**Comment by arm4b** [comment]
!bootstrap pack

**Comment by arm4b** [comment]
!bootstrap pack

**Comment by arm4b** [comment]
@dalarsen We're getting closer!
The new pack repo was bootstrapped.
Need a bit more help from you in the https://github.com/StackStorm-Exchange/stackstorm-openai/pull/1

**Comment by arm4b** [comment]
The pack is live!
https://exchange.stackstorm.org/#openai


@dalarsen Thanks for the contribution!
Please also accept Github invite so you'll have write access to the https://github.com/StackStorm-Exchange/stackstorm-openai repository.

**Pull request closed without merging**
